### PR TITLE
Implemented proper handling of ports with generations (like pyNN-xxx)

### PR DESCRIPTION
**Pull request by yurivict**
This needs to be reviewed and tested. It works for my original purpose.

The additional improvement is that now one can build the python port for non-default version, like this: net/py34-netifaces

### CODE REVIEW

**Comment by bdrewery** [comment]
Poudriere is not the right place for this, and the current idea of 'generations' is far too specific to python ports and the pkgname. It's an interesting idea but it has to be expanded to the bigger picture of subpackages and multiple packages per origin. Currently I won't take this as it's the wrong place to implement any of this. The framework needs to be mostly self-contained. The feature added in ports is just wrong.

File: src/share/poudriere/common.sh [hunk 1]
```diff
@@ -3105,16 +3126,61 @@ prefix_output() {
...
+}
+
+make_env_gen() {
+	local port="$1"
+	local gen="$2"
+	[ -z "${gen}" ] && return
+	case $(genport_extract_prefix $1) in
+		py) echo "PYTHON_VERSION=$(echo $gen | sed -E "s/(.)(.)/\1.\2/g")"
+		;;
+		ruby) echo "RUBY_VERSION=$gen"
```
Comment (bdrewery) [inline src/share/poudriere/common.sh]: This is fundamentally wrong. Poudriere should not have such specific handling here.

**Comment by bdrewery** [comment]
What we do need is to stop having Poudriere only track origins but to track pkgnames and relate those back to origins. We need a bigger picture solution to handle sub-packages and multi-version ports. All of which equate to Many Packages to 1 origin. I think you may actually be on the right track, but I don't like how specific to python this is.

**Comment by yurivict** [comment]
This is just a draft, because I hit this problem for the new port I needed. This of course needs to be expanded and generalized.

Port should return the set of it's generalization classes (make -VCLASSES), framework should know what values are allowed for each class. The idea, and 90% of this is expandable for this. You just picked the piece where python were mentioned. I just couldn't go the whole length in 6 hours that I spent.

**Comment by bdrewery** [comment]
What I would like to see is that a port defines some variable, such as your `CLASSES` idea or `GENERATIONS`, whatever. During the dependency lookup in `list_deps` we lookup this variable. If it has values then we act on those to generate virtual packages based on them. This may inject extra dependencies into the queue, such as in the python case to pass along this `DEPENDS_ARGS` flag (which I'd like to see changed). Poudriere would not need to be so strict about unique origins anymore. It should just build every package with the parameters stored with that package in the queue. When it builds it should just pass along `GENERATION=whatever` and have the framework handle the `DEPENDS_ARGS` etc.

**Comment by yurivict** [comment]
I will generalize this the way you suggested. The additional ideas are:
* Some serial port options (like DOCS, DEBUG, SOUND), or maybe even all port options can become "classes"
* Before the build, user can set the list of default class values (ex. DOCS=no, DOXYGEN=no) to the build tool (poudriere in this case), and such choices will be made for corresponding classes when they are available, but undefined. Choosing PYTHON=3.4 will effectively override the default python version.
* package names will reflect classes. For example, docsY-debugY-py31-netifaces may refer to the package for the port origin net/netifaces with DOCS=yes, DEBUG=yes, python version 31. (like this is now)
* class choices will be passed to the port make through the environment variable PORT_CLASSES. (instead of current PYTHON_VERSION)
* port infrastructure based on the class selection and its own logic will decide what dependencies it will have, and what classes dependencies will have. It will will return the list with class choices. (currently this is deduced from prefixNN- syntax for dependency). XXX_DEPENDS in port makefiles need to be generalized.

The current version implies one "class", "PYTHON", when the port name begins with py-, and no classes otherwise.

**Comment by bdrewery** [comment]
I keep making slave ports for my own needs. For example I have `mail/postfix` with _DOVECOT_ option. But I want a non-dovecot-depending version as well. So I added a `local/postfix-relay` as a slave port that I can set my own options on. The do conflict but that is fine in my case.

My need combined with the python needs seems like they could be solved in a unified way.

**Comment by yurivict** [comment]
This week I am going to commit the next phase that will add the build class, with many current binary options and python version becoming class parameters, so that poudriere will build with these options. Poudriere itself will only have the machinery, and no knowledge of particular classes. Yet next iteration should be concerned with how to pass class requirements  into dependencies in a nice way.

**Comment by koobs** [comment]
@yurivict Thank you for taking the initiative and your effort in working on this. It's probably worth introducing yourself to the rest of the FreeBSD / Python team on IRC (#freebsd-python @ freenode) so that we can help make progress :) cc @marcusva

Also see: #259

**Comment by bdrewery** [comment]
Thank you for your PR and patch to support it.  Since you sent it in though the dependency calculation code has been rewritten (last year) to improve parallelism.  Thus the patch was 70% no longer applicable. 

I implemented support for _DEPENDS_ARGS_ in #483, as well as _FLAVORS_.

**Pull request closed without merging**